### PR TITLE
 feat (api): implement Gitlab repository manager

### PR DESCRIPTION
**Pull request by proullon**

### CODE REVIEW

**Comment by jhayotte** [comment]
Can't wait to have this PR merged!

File: engine/api/repositoriesmanager/repogitlab/client.go [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+		URL:                   &url,
+		PushEvents:            &t,
+		MergeRequestsEvents:   &f,
+		TagPushEvents:         &f,
+		EnableSSLVerification: &f,
+	}
+
+	log.Debug("GitlabClient.CreateHook: %s %s\n", repo, *opt.URL)
+	_, _, err = c.client.Projects.AddProjectHook(repo, &opt)
+	if err != nil {
```
Comment (bnjjj) [inline engine/api/repositoriesmanager/repogitlab/client.go]: You can inline this

File: engine/api/repositoriesmanager/repogitlab/client.go [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+		Context:     &cds,
+		State:       getGitlabStateFromStatus(eventpb.Status),
+		Ref:         &eventpb.BranchName,
+		TargetURL:   &url,
+		Description: &desc,
+	}
+
+	_, _, err := c.client.Commits.SetCommitStatus(eventpb.RepositoryFullname, eventpb.Hash, opt)
+	if err != nil {
+		return err
```
Comment (bnjjj) [inline engine/api/repositoriesmanager/repogitlab/client.go]: Inline

File: engine/api/repositoriesmanager/repogitlab/driver.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+		if !ok {
+			return nil, fmt.Errorf("no app-id provided to Gitlab driver")
+		}
+		gd.ID = id
+
+		return gd, nil
+	}
+
+	err := json.Unmarshal([]byte(consumerData), &gd)
+	if err != nil {
```
Comment (bnjjj) [inline engine/api/repositoriesmanager/repogitlab/driver.go]: Inline

File: engine/api/hook.go [hunk 4]
```diff
@@ -21,6 +22,49 @@ import (
...
+
+	type gitlabEvent struct {
+		ObjectKind  string `json:"object_kind"`
+		Ref         string `json:"ref"`
+		UserName    string `json:"user_name"`
+		CheckoutSha string `json:"checkout_sha"`
+	}
+
+	var ge gitlabEvent
+	json.Unmarshal(data, &ge)
```
Comment (fsamin) [inline engine/api/hook.go]: check error and return if any

File: engine/api/repositoriesmanager/repogitlab/client.go [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+}
+
+//PullRequestEvents is not implemented
+func (c *GitlabClient) PullRequestEvents(string, []interface{}) ([]sdk.VCSPullRequestEvent, error) {
+	return nil, fmt.Errorf("Not implemented on Gitlab")
+}
+
+// Release not implemented in Gitlab
+func (c *GitlabClient) Release(repo string, tagName string, title string, releaseNote string) (*sdk.VCSRelease, error) {
+	return nil, fmt.Errorf("Gitlab do not provide release system")
```
Comment (sguiheux) [inline engine/api/repositoriesmanager/repogitlab/client.go]: https://docs.gitlab.com/ee/api/tags.html#create-a-new-release

Comment (proullon) [reply engine/api/repositoriesmanager/repogitlab/client.go]: yeah I known haha, I'll change the error message, but I'll implement it later

**Comment by proullon** [comment]
done

I'm not comfortable with inlined error though, it makes code harder to read since the function call is clogged with error management characters whereas the eye is used to ignore the usual error management:
```go
if err != nil {
   return err
}
```
Besides you can't always inline since you may have variables you want to keep out of error scope, so it's not possible to do this every time, then the code structure is eventually inconsistent...

Anyway, why do you want to inline errors ?

**Comment by yesnault** [comment]
yes, we inline, when it's possible, the err management.
By inlining all existing err management, we discover so many errors not catched because of var err is used everywhere.

```
err:= aFunc()
if err != nil {
// err management
}

err:= anotherFunc()
//other code here, Oops, we forget manage err

err:= anotherFunc()
if err != nil {
// err management
}
```

By inlining, this avoid a lot of unchecked errors.

**Comment by yesnault** [comment]
and a big thank you for this PR!

**Comment by proullon** [comment]
I mean, errcheck ? https://github.com/kisielk/errcheck

**Comment by yesnault** [comment]
why not, but I'm not sure that errcheck can detect  that anotherFunc2 is not check, because of err is used and check, after the call to anotherFunc3 (and this is the real issue: err affected, but not checked directly before being re-affected after).


```golang
err:= aFunc()
if err != nil {
// err management
}

err:= anotherFunc2()
//other code here, Oops, we forget manage err

err:= anotherFunc3()
if err != nil {
// err management
}
```

**Pull request merged**